source/ada/lsp-ada_handlers.adb

```diff
@@ -431,14 +434,206 @@ package body LSP.Ada_Handlers is
+           (Subp_Spec_Node.Text);
+         Lines : GNAT.Strings.String_List_Access := Split
+           (Text,
+            On               => ASCII.LF,
+            Omit_Empty_Lines => True);
+         Idx   : Integer;
+      begin
+         --  For single-line subprogram specifications, we display the
+         --  associated text directly.
+         --  For multi-line ones, remove the identation blankspaces to replace
+         --  them by a fixed number of blankspaces.
+
+         if Lines'Length = 1 then
+            Hover_Text := To_LSP_String (Text);
+         else
+            Hover_Text := To_LSP_String (Lines (Lines'First).all);
+
+            for J in Lines'First + 1 .. Lines'Last loop
+               Idx := Lines (J)'First;
+               Skip_Blanks (Lines (J).all, Idx);
+
+               Hover_Text := Hover_Text
+                 & To_LSP_String
+                 (ASCII.LF
+                  & (if Lines (J).all (Idx) = '(' then "  " else "   ")
+                  & Lines (J).all (Idx .. Lines (J).all'Last));
+            end loop;
+         end if;
+
+         GNAT.Strings.Free (Lines);
+      end Create_Hover_Text_For_Subp_Spec;
+
    begin
-      Response.error :=
-        (True,
-         (code => LSP.Messages.InternalError,
-          message => To_LSP_String ("Not implemented"),
-          data => <>));
+      if Name_Node = No_Name then
+         return Response;
+      end if;
+
+      --  Get the defining node of the node being hovered
+      Defining_Name_Node := LSP.Lal_Utils.Resolve_Name
+        (Name_Node => Name_Node);
+
+      if Defining_Name_Node = No_Defining_Name then
+         return Response;
+      end if;
+
+      --  Get the associated basic declaration
+      Decl := Defining_Name_Node.P_Basic_Decl;
+
+      if Decl = No_Basic_Decl then
+         return Response;
+      end if;
+
+      --  Try to retrieve the subprogram spec node, if any: if it's a
+      --  subprogram node that does not have any separate declaration we
+      --  only want to display its specification, not the body.
+      Subp_Spec_Node := Decl.P_Subp_Spec_Or_Null;
+
+      if Subp_Spec_Node /= No_Base_Subp_Spec then
+         Create_Hover_Text_For_Subp_Spec;
+      else
+         Create_Hover_Text_For_Basic_Decl;
+      end if;
+
+      --  Append the associated basic declaration's text to the response.
+      --  We set the language for highlighting.
+      if Hover_Text /= Empty_LSP_String then
+         Response.result.contents.Append
+           (LSP.Messages.MarkedString'
+              (Is_String => False,
+               value     => Hover_Text,
+               language  => To_LSP_String ("ada")));
+
+         --  TODO: append the trivias associated with the basic declaration
+         --  node when available in libadalang.
+         Response.result.contents.Append
+           (LSP.Messages.MarkedString'
+              (Is_String => True,
+               value     => To_LSP_String ("TODO: append the comments")));
```

Comment:
I wouldn't leave the "TODO" string visible to clients

Reply:
Well the trace that enables LSP based tooltips won't be enabled by default so they won't see it.